controllers/managedocs_controller.go

```diff
@@ -175,14 +175,23 @@ func (r *ManagedOCSReconciler) generateStorageCluster(
 	if reconcileStrategy == v1.ReconcileStrategyStrict {
 		// Get an instance of the desired state
 		desired := utils.ObjectFromTemplate(templates.StorageClusterTemplate, r.Scheme).(*ocsv1.StorageCluster)
-
 		if err := r.setStorageClusterParamsFromSecret(desired); err != nil {
 			return err
 		}
 		// Override storage cluster spec with desired spec from the template.
 		// We do not replace meta or status on purpose
 
-		sc.Spec = desired.Spec
+		if sc.Spec.StorageDeviceSets == nil {
+			sc.Spec = desired.Spec
+		} else {
+			if sc.Spec.StorageDeviceSets[0].Count <= desired.Spec.StorageDeviceSets[0].Count {
+				sc.Spec = desired.Spec
+				r.Log.Info("Actual storage cluster parameter", "actualCount", sc.Spec.StorageDeviceSets[0].Count)
+			} else {
+				err := fmt.Errorf("cannot set storage device set count: actualCount=%v", sc.Spec.StorageDeviceSets[0].Count)
+				r.Log.Error(err, "Reducing storage device set count is not allowed")
+			}
+		}
```

Comment:
Is there a reason this entire block should reside here and not where the actual setting of the count is?
By doing it this way you split the logic of setting the count into 2 distinct places and, which is harder to reason about, as will force you to reimplement some of the logic that is already implemented inside the `setStorageClusterParamsFromSecret` func (like finding the device set by name)